### DISCUSSION TITLE
Unable to run Qodana workflow with Java 20

### DISCUSSION THREAD

**Discussion by rob93c** [JetBrains/Qodana#181] [Q&A] [answered]
Hi all! 
I'm setting up Qodana for my project with [this pull request](https://github.com/Stickerifier/Stickerify/pull/129/files)

I had a [successful run](https://github.com/Stickerifier/Stickerify/actions/runs/5706248124/job/15461771398#step:4:12732) that showed the following message:

> request to https://api.qodana.cloud/v1/reports/ failed with status code: 404, exception: no-token, reason: There is no token with id A6dEn

I then tried to retrigger the job but it kept failing with this error ([a sample failing run](https://github.com/Stickerifier/Stickerify/actions/runs/5706459781/job/15462160022#step:4:72)):

> Please, specify sdk 'temurin-20' for module 'Stickerify.main'
2023-07-30 13:19:03,893 [  39958] SEVERE - #o.j.q.s.i.r.QodanaInspectionApplication - Cannot configure project to run inspections

Did I misconfigured anything?

**Comment by brichbash** [comment]
Hello,
Could you please try setting `temurin-20` in `projectJDK` in your `qodana.yaml`?
E.g.:
```
version: "1.0"
profile:
  name: qodana.recommended
linter: jetbrains/qodana-jvm:latest
projectJDK: temurin-20
```

**Reply by rob93c** [reply]
Hi @brichbash and thanks for helping me figure this out!
Unfortunately, the change still results in the [same error](https://github.com/Stickerifier/Stickerify/actions/runs/5753376829/job/15596436225#step:4:69):

> Please, specify sdk 'temurin-20' for module 'Stickerify.main'
2023-08-03 16:38:31,359 [  39727] SEVERE - #o.j.q.s.i.r.QodanaInspectionApplication - Cannot configure project to run inspections
org.jetbrains.qodana.staticAnalysis.inspections.runner.QodanaException: Cannot configure project to run inspections

**Reply by brichbash** [reply, marked as answer]
Please try removing all qodana-related caches from https://github.com/Stickerifier/Stickerify/actions/caches/. If it's still reproduced, set `upload-result` to `true` in [your Qodana job configuration file](https://github.com/Stickerifier/Stickerify/blob/configure-qodana-analysis/.github/workflows/qodana-code-quality.yml#L30) and run Qodana again:
```
with:
  args: -v,${{ env.JAVA_HOME }}:/root/.jdks/jdk
  cache-default-branch-only: true
  post-pr-comment: false
  push-fixes: none
  upload-result: true
```
This way, we'll be able to retrieve logs from the Qodana artifact

**Reply by rob93c** [reply]
The cache was indeed the issue here! I deleted the relevant one and the build worked properly (uploading the report too).
I then proceeded adding to my workflow `use-caches: false` to disable it but the [subsequent action run](https://github.com/Stickerifier/Stickerify/actions/runs/5753893591/job/15598030745#step:4:55) showed the error:

> 2023-08-03 17:34:13,102 [  39508] SEVERE - #o.j.q.s.i.r.QodanaInspectionApplication - Can't find locally or download required in config JDK 'temurin-20'
org.jetbrains.qodana.staticAnalysis.inspections.runner.QodanaException: Can't find locally or download required in config JDK 'temurin-20'

I then removed the explicit `projectJDK` from my `qodana.yaml` and it worked just fine 🎆 

Just before I mark this question as answered, do you foresee any issue with the changes I applied?

**EDIT:** I just noticed that Qodana is marking as errors Java 20 preview features used by the project, as per [this discussion](https://github.com/JetBrains/Qodana/discussions/129#discussioncomment-3981026) I committed my `compiler.xml` file:
````xml
<?xml version="1.0" encoding="UTF-8"?>
<project version="4">
  <component name="CompilerConfiguration">
    <bytecodeTargetLevel target="20" />
  </component>
  <component name="JavacSettings">
    <option name="ADDITIONAL_OPTIONS_STRING" value="--enable-preview" />
    <option name="ADDITIONAL_OPTIONS_OVERRIDE">
      <module name="Stickerify.main" options="--add-exports java.desktop/sun.font=ALL-UNNAMED --add-exports java.desktop/sun.java2d.pipe=ALL-UNNAMED --add-exports java.desktop/sun.awt=ALL-UNNAMED --add-exports java.desktop/sun.awt.X11=ALL-UNNAMED --add-exports java.desktop/sun.awt.datatransfer=ALL-UNNAMED --add-exports java.desktop/com.sun.imageio.plugins.jpeg=ALL-UNNAMED --add-exports java.desktop/sun.awt.im=ALL-UNNAMED --add-exports java.desktop/sun.awt.image=ALL-UNNAMED --add-exports java.desktop/sun.java2d=ALL-UNNAMED --add-exports java.desktop/sun.java2d.loops=ALL-UNNAMED --add-exports java.desktop/sun.java2d.opengl=ALL-UNNAMED --add-exports java.desktop/sun.swing=ALL-UNNAMED" />
    </option>
  </component>
</project>
````
Despite adding this file, the error is still reported by the tool: https://qodana.cloud/projects/3XQOZ/reports/PgbDd?showingBaseline=GYQwNgzgpkA#245cb07cdcc7a0fdef7fbb7337fba8c52c8aa742eae2591253b6d0a2ffa9857c

Thank you for your precious help

**Reply by brichbash** [reply]
Could you please share your logs by enabling `upload-result` the way I described in the previous comment? As for `projectJDK`. I didn't notice you were mounting your JDK in the container. In this case `projectJDK` is not needed.

**Reply by rob93c** [reply]
> I didn't notice you were mounting your JDK in the container.

Sorry if I didn't make it clear, I had to since Qodana doesn't support Java 20 out of the box yet.

I just enabled artifact upload, here's it: https://github.com/Stickerifier/Stickerify/suites/14819521515/artifacts/843617105

**Reply by rob93c** [reply]
Hi @brichbash, do you have any update on this regard?

**Reply by brichbash** [reply]
Hi, sorry for the delay, I was on vacation.

I noticed some strange behaviour with `projectJDK` field in `qodana.yaml` while trying to reproduce your problem.

Could you try removing the step with downloading JDK in Qodana job configuration and also the `args: -v,${{ env.JAVA_HOME }}:/root/.jdks/jdk` part? After that, commit or add with Qodana bootstrap `.idea/misc.xml`:
```
<?xml version="1.0" encoding="UTF-8"?>
<project version="4">
  <component name="ExternalStorageConfigurationManager" enabled="true" />
  <component name="ProjectRootManager" version="2" languageLevel="JDK_20_PREVIEW" project-jdk-name="temurin-20" project-jdk-type="JavaSDK" />
</project>
```
To add it with bootstrap without committing the file to the repo add to `qodana.yaml`:
```
bootstrap: |
  mkdir .idea
  cat > .idea/misc.xml << \EOF
  <?xml version="1.0" encoding="UTF-8"?>
  <project version="4">
  <component name="ExternalStorageConfigurationManager" enabled="true" />
  <component name="ProjectRootManager" version="2" languageLevel="JDK_20_PREVIEW" project-jdk-name="temurin-20" project-jdk-type="JavaSDK" />
  </project>
  EOF
```
For some reason, this way Qodana is able to resolve `temurin-20` and download it.

**Reply by rob93c** [reply]
Please do not worry and thanks as usual for your help!

I removed the `compiler.xml` file and [I applied your suggestion](https://github.com/Stickerifier/Stickerify/pull/129/commits/fdbb19c174ae5679866d3ac2c4bb01485dc552ed) but, although the builds still work, Qodana keeps reporting those errors:
![image](https://github.com/JetBrains/Qodana/assets/1816483/aef3f352-84f8-47f1-96ff-fd5e54bb5c64)

**Reply by brichbash** [reply]
Ok, I believe I found the root cause. Using Java Preview features requires accepting `terms of legal notice of the beta Java specification`, which is implemented by a dialog in IntelliJ IDEA, and in Qodana it’s not possible at the moment to accept the notice. In the `EAP` builds the agreement is not required, and that’s why I was able to make it work. Our internal images don’t require accepting it.
If you have a spare minute, please vote for the issue[ on our tracker](https://youtrack.jetbrains.com/issue/QD-6990/Support-using-Java-preview-features) to get notified of updates and help us prioritize the problem.

**Reply by rob93c** [reply]
Oh, that explains it! Great job with this triage, I just voted for the issue.
It isn't a blocking problem on my side so I'll enable Qodana on my project despite those 4 alerts

Let me thank you again for all your support.

Have a great day,
Roberto

**Reply by brichbash** [reply]
@rob93c We've published EAP linters today (their license will expire in 60 days). Try using `jetbrains/qodana-jvm:2023.2-eap` instead with the following [yaml](https://github.com/brichbash/Stickerify/blob/configure-qodana-analysis/qodana.yaml). Java with Preview features is enabled with this image (Checked on [this fork](https://github.com/brichbash/Stickerify/actions/runs/6086690104)).

**Reply by rob93c** [reply]
Please, feel free to raise a pull request against the project if you'd like to contribute otherwise I'll do it myself in the coming days

**Reply by rob93c** [reply]
Just a quick update @brichbash: I added the changes you suggested and everything works as expected now 🔥
